src/runner.py

```diff
@@ -171,47 +158,54 @@ def get_info(self):
 
     def reconcile(self, virt_type, quantity):
         """Bring runners in line with target."""
-        active_runners = []
-        stale_runners = []
-        for runner in self.get_info():
-            if runner.virt_type == virt_type:
-                if runner.is_active:
-                    active_runners.append(runner)
-                else:
-                    stale_runners.append(runner)
+        runners = [r for r in self.get_info() if r.virt_type == virt_type]
+        online_runners = [r for r in runners if r.is_online and r.is_local]
+        offline_runners = [r for r in runners if r.is_offline]                
 
-        if stale_runners:
-            runner_names = ", ".join(r.name for r in stale_runners)
+        # Clean up stale (offline) runners
+        if offline_runners:
+            runner_names = ", ".join(r.name for r in offline_runners)
             logger.info(f"Cleaning up stale {virt_type} runners: {runner_names}")
-            for runner in stale_runners:
+            for runner in offline_runners:
                 self._remove_runner(runner)
-
-        delta = quantity - len(active_runners)
+        
+        # Add/Remove runners to match the target quantity
+        delta = quantity - len(online_runners)
         if delta > 0:
             logger.info(f"Adding {delta} additional {virt_type} runners")
             for i in range(delta):
                 self.create(image="ubuntu", virt=virt_type)
         elif delta < 0:
-            active_runners.sort(key=lambda r: r.created_at)
-            old_runners = active_runners[abs(delta) :]
+            online_runners.sort(key=lambda r: r.local.created_at)
+            old_runners = online_runners[abs(delta) :]
             runner_names = ", ".join(r.name for r in old_runners)
             logger.info(f"Removing extra {virt_type} runners: {runner_names}")
             for runner in old_runners:
                 self._remove_runner(runner)
         return delta
 
     def clear(self):
-        """Clear out all existing runners."""
-        runners = self.get_info()
+        """Clear out existing local runners."""
+        runners = [r for r in self.get_info() if r.is_local]
         runner_names = ", ".join(r.name for r in runners)
-        logger.info(f"Removing existing runners: {runner_names}")
+        logger.info(f"Removing existing local runners: {runner_names}")
         for runner in runners:
             self._remove_runner(runner)
 
     def create(self, image, virt="container"):
         """Create a runner"""
         instance = self._create_instance(image=image, virt=virt)
         instance.start(wait=True)
+
+        # Wait for the instance to boot
+        for attempt in range(8):
+            try:
+                self._check_output(instance, "who")
```

Comment:
The call to `instance.start(wait=True)` returns as soon as the instance started, which does not mean that it has fully booted up and is ready to execute commands. Let's run a few wait iterations to give the instance time to boot up.

Reply:
Looking at this from the perspective of someone using the charm and has access to a "slow" cloud, then failing the deployment because of a too short deadline will be a showstopper. Catching it in a try-except block will not solve it, nor does the 2 minute interval. At the same time, it is not a good idea to wait indefinitely.
What about making the timeout configurable with a default of 2 minutes and wrapping the wait in a try-except block?

Reply:
I used the `reconcile-interval` config value as the upper limit for how long to wait for an instance to start. It gives operators the ability to configure the `reconcile-interval` such that it is long enough for at least one runner to start. At the same time it "motivates" the operator to use a sufficiently long interval to ensure stability based on the environment.